### PR TITLE
Feature/#7 create explode method

### PR DESCRIPTION
**Pull request by mserier**

### CODE REVIEW

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 1]
```diff
@@ -4,16 +4,84 @@
...
+    {
+        
+    }
+    
     void Update()
     {
         
     }
+    
+    private Vector3[] _calculatedPathPoints = null;
```
Comment (Maxboef) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]: Move to top of class

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 2]
```diff
@@ -4,16 +4,84 @@
...
+    }
+
+    private static bool _initializing = false;
+    public static void Initialize()
     {
+        if (_instance == null && _initializing == false)
+        {
+            _initializing = true;
+
+            GameManager instance = FindObjectOfType<GameManager>();
```
Comment (Maxboef) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]:
```suggestion
            var instance = FindObjectOfType<GameManager>();
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 3]
```diff
@@ -5,30 +5,84 @@
...
 
     void Start()
     {
+        _energyCharge = EnergyCapacity;
+        _rigidbody = GetComponent<Rigidbody>();
+    }
+    
+    private void Update()
+    {
+        if (_waypointIndex < GameManager.Instance.GetLevelPath().Length)
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]: This if should be inverted

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 4]
```diff
@@ -5,30 +5,84 @@
...
+                //Move enemy parallel to the path
+                _rigidbody.AddForce(_rigidbody.mass * (MovementSpeed * (GameManager.Instance.GetLevelPath()[_waypointIndex] - GameManager.Instance.GetLevelPath()[_waypointIndex - _lookAheadDistance]).normalized));
+            }
+        }
     }
 
     void FixedUpdate()
     {
-        if (_energy < 0)
+        //Calculate energy potential
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]:
```suggestion
        // Calculate energy potential
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 5]
```diff
@@ -5,30 +5,84 @@
...
+            {
+                //Lost so go back to last known waypoint
+                _rigidbody.AddForce(_rigidbody.mass * (MovementSpeed * 2 * (GameManager.Instance.GetLevelPath()[_waypointIndex] - transform.position).normalized));
+            }
+            else
+            {
+                //Move enemy towards path
+                _rigidbody.AddForce(_rigidbody.mass * (MovementSpeed * _potentialEnergy * (GameManager.Instance.GetLevelPath()[_waypointIndex + _lookAheadDistance] - transform.position).normalized));
 
+                //Move enemy parallel to the path
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]:
```suggestion
                // Move enemy parallel to the path
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 6]
```diff
@@ -5,30 +5,84 @@
...
+            }
+
+            if(_lost)
+            {
+                //Lost so go back to last known waypoint
+                _rigidbody.AddForce(_rigidbody.mass * (MovementSpeed * 2 * (GameManager.Instance.GetLevelPath()[_waypointIndex] - transform.position).normalized));
+            }
+            else
+            {
+                //Move enemy towards path
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]:
```suggestion
                // Move enemy towards path
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 7]
```diff
@@ -69,12 +142,61 @@ public void Explode()
...
+        Credit.Value = Value;
+        Instantiate(Credit, gameObject.transform.position, gameObject.transform.rotation);
+    }
+
+    /// <summary>
+    /// This function removes the enemy and starts the teleport particle system and sound effect
+    /// </summary>
+    public void Finish()
+    {
+        //if not in the world, instantiate
```
Comment (Maxboef) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]:
```suggestion
        // If not in the world, instantiate
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 8]
```diff
@@ -5,30 +5,84 @@
...
+
+            //Following if statement isn't needed if we spawn the enemy in the correct place
+            if (_waypointIndex == 0)
+            {
+                transform.position = GameManager.Instance.GetLevelPath()[_waypointIndex + _lookAheadDistance];
+            }
+
+            if(_lost)
+            {
+                //Lost so go back to last known waypoint
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]:
```suggestion
                // Lost so go back to last known waypoint
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 9]
```diff
@@ -5,30 +5,84 @@
...
+        _rigidbody = GetComponent<Rigidbody>();
+    }
+    
+    private void Update()
+    {
+        if (_waypointIndex < GameManager.Instance.GetLevelPath().Length)
+        {
+            float step = MovementSpeed * Time.deltaTime;
+
+            //Following if statement isn't needed if we spawn the enemy in the correct place
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]:
```suggestion
            // Following if statement isn't needed if we spawn the enemy in the correct place
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 10]
```diff
@@ -5,30 +5,84 @@
...
     {
+        _energyCharge = EnergyCapacity;
+        _rigidbody = GetComponent<Rigidbody>();
+    }
+    
+    private void Update()
+    {
+        if (_waypointIndex < GameManager.Instance.GetLevelPath().Length)
+        {
+            float step = MovementSpeed * Time.deltaTime;
```
Comment (Maxboef) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]:
```suggestion
            var step = MovementSpeed * Time.deltaTime;
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 11]
```diff
@@ -5,30 +5,84 @@
...
-    private float _energy = 100f;
+    private float _potentialEnergy = 1f;
+    private int _lookAheadDistance = 5;//(in waypoints)
 
     public ParticleSystem ExplodeEffect;
     private ParticleSystem _explodeEffectInstance = null;
+    public ParticleSystem TeleportEffect;
+    private ParticleSystem _teleportEffectInstance = null;
+    public Credit Credit;
+    public int Value;
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]: This property should be renamed to `CreditValue`.

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 12]
```diff
@@ -4,16 +4,84 @@
...
+            if (_instance == null)
+            {
+                Initialize();
+            }
+
+            return _instance;
+        }
+    }
+
+    private static bool _initializing = false;
```
Comment (Maxboef) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]: To top of Class

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 13]
```diff
@@ -60,7 +114,26 @@ public void Heal(float healAmount)
...
+    public void Charge(float chargeAmount)
+    {
+        _energyCharge = Mathf.Clamp(_energyCharge + chargeAmount, 0, EnergyCapacity);
+    }
+
+    /// <summary>
+    /// This function will remove the specified chargeAmount from the enemy's charge
+    /// </summary>
+    /// <param name="chargeAmount"></param>
+    public void DisCharge(float chargeAmount)
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]:
```suggestion
    public void Discharge(float amount)
```

File: src/Assets/LightSkyDefense/Scripts/Path.cs [hunk 14]
```diff
@@ -13,7 +13,15 @@ public class Path : MonoBehaviour
 
     void Start()
     {
+        //Get the path points, convert the coordinates to world space and save it in the GameManager
         var pathVectors = GetVector3sCoordinatesFromPath(LineDivision);
+        var pathWorldVectors = new Vector3[pathVectors.Length];
+        for(int i=0;i<pathVectors.Length;i++)
```
Comment (Maxboef) [inline src/Assets/LightSkyDefense/Scripts/Path.cs]:
```suggestion
        for(int i=0; i<pathVectors.Length; i++)
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 15]
```diff
@@ -69,12 +142,61 @@ public void Explode()
...
+        Destroy(gameObject);
+    }
+
+    /// <summary>
+    /// When the enemy collides with a waypoint collider, update the waypoint index
+    /// </summary>
+    /// <param name="col"></param>
+    private void OnTriggerEnter(Collider col)
+    {
+        if(col.gameObject.name.StartsWith("Way"))
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]: If statement should be inverted

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 16]
```diff
@@ -4,16 +4,84 @@
...
+    public static void Initialize()
     {
+        if (_instance == null && _initializing == false)
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]: If statement should be inverted

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 17]
```diff
@@ -4,16 +4,84 @@
...
+
+            _instance = instance;
+   
+            _initializing = false;
+        }
     }
 
-    // Update is called once per frame
+    
+    void Start()
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]: Empty methods should be removed.

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 18]
```diff
@@ -4,16 +4,84 @@
...
+    {
+        
+    }
+    
     void Update()
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]: Empty methods should be removed.

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 19]
```diff
@@ -4,16 +4,84 @@
...
     {
         
     }
 
+    public Vector3[] GetLevelPath()
+    {
+        if(_calculatedPathPoints==null)
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]: Can be a ternary expression.
```
return _calculatedPathPoints == null 
    ? new Vector3[0]
    : _calculatedPathPoints;
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 20]
```diff
@@ -5,30 +5,84 @@
...
+                _rigidbody.AddForce(_rigidbody.mass * (MovementSpeed * (GameManager.Instance.GetLevelPath()[_waypointIndex] - GameManager.Instance.GetLevelPath()[_waypointIndex - _lookAheadDistance]).normalized));
+            }
+        }
     }
 
     void FixedUpdate()
     {
-        if (_energy < 0)
+        //Calculate energy potential
+        _potentialEnergy = 0.8f-Vector3.Distance(transform.position, GameManager.Instance.GetLevelPath()[_waypointIndex]);
```
Comment (Maxboef) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]: magic number?

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 21]
```diff
@@ -4,16 +4,84 @@
...
+    }
+
+    public void SetPathPoints(Vector3[] pathPoints, GameObject parent)
+    {
+        _calculatedPathPoints = pathPoints;
+
+        //Spawn road waypoints
+        //Quaternion quaternion = new Quaternion();
+        int i = 0;
+        foreach (Vector3 pathPoint in GetLevelPath())
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]: Convert this to a for loop if you need the iterator index.

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs (hunk 20 above)
Comment (Maxboef) [reply src/Assets/LightSkyDefense/Scripts/Enemy.cs]: 0.8f

File: src/Assets/LightSkyDefense/Scripts/Path.cs (hunk 14 above)
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Path.cs]:
```suggestion
        for(int i=0; i < pathVectors.Length; i++)
```

File: src/Assets/LightSkyDefense/Scripts/Path.cs [hunk 22]
```diff
@@ -13,7 +13,15 @@ public class Path : MonoBehaviour
...
+        {
+            pathWorldVectors[i] = transform.TransformPoint(pathVectors[i]);
+        }
+        GameManager.Instance.SetPathPoints(pathWorldVectors, gameObject);
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Path.cs]: New lines around new scopes.

```suggestion

        GameManager.Instance.SetPathPoints(pathWorldVectors, gameObject);
```

File: src/Assets/LightSkyDefense/Scripts/Path.cs (hunk 14 above)
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Path.cs]: New lines around new scopes.

```suggestion

        for(int i=0;i<pathVectors.Length;i++)
```

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 23]
```diff
@@ -4,16 +4,84 @@
...
+        {
+            return _calculatedPathPoints;
+        }
+    }
+
+    public void SetPathPoints(Vector3[] pathPoints, GameObject parent)
+    {
+        _calculatedPathPoints = pathPoints;
+
+        //Spawn road waypoints
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]:
```suggestion
        // Spawn road waypoints
```

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 24]
```diff
@@ -4,16 +4,84 @@
...
+        //Quaternion quaternion = new Quaternion();
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]:
```suggestion
```

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 25]
```diff
@@ -4,16 +4,84 @@
 
 public class GameManager : MonoBehaviour
 {
-    // Start is called before the first frame update
-    void Start()
+    [HideInInspector]
+    public GameObject _wayPointPrefab;
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]: I think we could just set this value in scene, setting the value of the `_wayPointPrefab` using a script feels like a hack to me.

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 26]
```diff
@@ -4,16 +4,84 @@
...
+
+    private Vector3[] _calculatedPathPoints = null;
+
+    private static GameManager _instance;
+    public static GameManager Instance
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]: It's not necessary to make this object a singleton, the game manager will be part of the Player object. the Player object is already singleton. We can just use `Player.instance.GetComponent<GameManager>()` to get the GameManager instance.

File: src/Assets/LightSkyDefense/Scripts/GameManager.cs [hunk 27]
```diff
@@ -4,16 +4,84 @@
...
+        {
+            return new Vector3[0];
+        }
+        else
+        {
+            return _calculatedPathPoints;
+        }
+    }
+
+    public void SetPathPoints(Vector3[] pathPoints, GameObject parent)
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/GameManager.cs]: I think this should be part of the `Path` and not the `GameManager`.

File: src/Assets/LightSkyDefense/Scripts/Path.cs [hunk 28]
```diff
@@ -13,7 +13,15 @@ public class Path : MonoBehaviour
 
     void Start()
     {
+        //Get the path points, convert the coordinates to world space and save it in the GameManager
         var pathVectors = GetVector3sCoordinatesFromPath(LineDivision);
+        var pathWorldVectors = new Vector3[pathVectors.Length];
+        for(int i=0; i<pathVectors.Length; i++)
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/Path.cs]:
```suggestion
        for(var i = 0; i < pathVectors.Length; i++)
```

File: src/Assets/LightSkyDefense/Scripts/Enemy.cs [hunk 29]
```diff
@@ -1,34 +1,90 @@
...
+            }
+        }
+
+        if (_energyCharge < 0)
         {
             Damage(30);
         }
 
-        //todo: intergrate  behaviour
+        //
```
Comment (Maxboef) [inline src/Assets/LightSkyDefense/Scripts/Enemy.cs]:
```suggestion
       
```

**Pull request merged**